File: html/js/tunstall.js
/*
Nexus

Copyright(C) 2012 - Federico Ponchio
ISTI - Italian National Research Council - Visual Computing Lab

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License (http://www.gnu.org/licenses/gpl.txt)
for more details.
*/

BitStream = function(array) {
	var t = this;
	t.a = array;
	t.current = array[0];
	t.position = 0; //position in the buffer
	t.pending = 32;  //bits still to read
	t.mask = new Uint32Array([0x00, 0x01, 0x03, 0x07, 0x0f, 0x01f, 0x03f, 0x07f, 0xff, 
		0x01ff, 0x03ff, 0x07ff, 0x0fff, 0x01fff, 0x03fff, 0x07fff, 0xffff, 0x01ffff,
		0x03ffff, 0x07ffff, 0x0fffff, 0x01fffff, 0x03fffff, 0x07fffff, 0xffffff, 0x01ffffff,
		0x03ffffff,  0x07ffffff,  0x0fffffff, 0x01fffffff, 0x03fffffff, 0x7fffffff]);

};
 
BitStream.prototype = { 
	read: function(bits) {
		var t = this;
		var result = 0;
		if(bits > t.pending) {
			result |= (t.current << (bits - t.pending))>>>0;
			bits -= t.pending;
			t.current = t.a[++t.position];
			t.pending = 32;
		}
		t.pending -= bits;
		result |= (t.current >>> t.pending);
		t.current = (t.current & t.mask[t.pending])>>>0;
		return result;
	}
};

Stream = function(buffer) {
	var t = this;
	t.data = buffer;
	t.buffer = new Uint8Array(buffer);
	t.pos = 0;
	t.view = new DataView(buffer);
	t.max = new Uint32Array(32);
	for(var i = 0; i < 32; i++)
		t.max[i] = (1<<i)>>>1;
}

Stream.prototype = {
	readChar: function() {
		var c = this.buffer[this.pos++];
		if(c > 127) c -= 256;
		return c;
	},
	readUChar: function() {
		return this.buffer[this.pos++];
	},	
	readShort: function() {
		this.pos += 2;
		return this.view.getInt16(this.pos-2, true);
	},
	readFloat: function() {
		this.pos += 4;
		return this.view.getFloat32(this.pos-4, true);
	},
	readInt: function() {
		this.pos += 4;
		return this.view.getInt32(this.pos-4, true);
	},
	readArray: function(n) {
		var a = this.buffer.subarray(this.pos, this.pos+n);
		this.pos += n;
		return a;
	},
	readString: function() {
		var n = this.readShort();
		var s = String.fromCharCode.apply(null, this.readArray(n-1));
		this.pos++; //null terminator of string.
		return s;
	},
	readBitStream:function() {
		var n = this.readInt();
		var pad = this.pos & 0x3;
		if(pad != 0)
			this.pos += 4 - pad;
		var b = new BitStream(new Uint32Array(this.data, this.pos, n));
		this.pos += n*4;
		return b;
	},
	//make decodearray2,3 later //TODO faster to create values here or passing them?
	decodeArray: function(N, values) {
		var t = this;
		var bitstream = t.readBitStream();

		var tunstall = new Tunstall;
		var logs = tunstall.decompress(this);

		for(var i =0; i < logs.length; i++) {
			var diff = logs[i];
			if(diff == 0) {
				for(var c = 0; c < N; c++)
					values[i*N + c] = 0;
				continue;
			}
//			var max = t.max[diff];
			var max = (1<<diff)>>>1;
			for(var c = 0; c < N; c++)
				values[i*N + c] = bitstream.read(diff) - max;
		}
		return logs.length;
	},

	decodeValues: function(N, values) {
		var t = this;
		var bitstream = t.readBitStream();
		var tunstall = new Tunstall;

		for(var c = 0; c < N; c++) {
			var logs = tunstall.decompress(this);

			for(var i = 0; i < logs.length; i++) {
				var diff = logs[i];
				if(diff == 0) {
					values[i*N + c] = 0;
					continue;
				}

				var val = bitstream.read(diff);
				var middle = (1<<(diff-1))>>>0;
				if(val < middle)
					val = -val -middle;
				values[i*N + c] = val;
			}
		}
		return logs.lenght;
	}
};



function Tunstall(wordsize, lookup_size) {
	this.wordsize = wordsize? wordsize : 8;
	this.lookup_size = lookup_size? lookup_size : 8;
}

/*function Tunstall() {
}

Tunstall.prototype.wordsize = 8;
Tunstall.prototype.lookup_size = 8; */



Tunstall.prototype = {
	decompress: function(stream) {
		var nsymbols = stream.readUChar();
		this.probs = stream.readArray(nsymbols*2);
		this.createDecodingTables2();
		var size = stream.readInt();
		if(size > 100000000) throw("TOO LARGE!");
		var data = new Uint8Array(size);
		var compressed_size = stream.readInt();
		if(size > 100000000) throw("TOO LARGE!");
		var compressed_data = stream.readArray(compressed_size);
		if(size)
			this._decompress(compressed_data, compressed_size, data, size);
		return data;
	}, 

	createDecodingTables: function() {
		//read symbol,prob,symbol,prob as uchar.
		//Here probs will range from 0 to 0xffff for better precision

		var n_symbols = this.probs.length/2;
		if(n_symbols <= 1) return;

		var queues = []; //array of arrays
		var buffer = []; 


		//initialize adding all symbols to queues
		for(var i = 0; i < n_symbols; i++) {
			var symbol = this.probs[i*2];
			var s = [(this.probs[i*2+1])<<8, buffer.length, 1]; //probability, position in the buffer, length
			queues[i] = [s];
			buffer.push(this.probs[i*2]); //symbol
		}
		var dictionary_size = 1<<this.wordsize;
		var n_words = n_symbols;
		var table_length = n_symbols;

		//at each step we grow all queues using the most probable sequence
		while(n_words < dictionary_size - n_symbols +1) {
			//Should use a stack or something to be faster, but we have few symbols
			//find highest probability word
			var best = 0;
			var max_prob = 0;
			for(var i = 0; i < n_symbols; i++) {
				var p = queues[i][0][0]; //front of queue probability.
				if(p > max_prob) {
					best = i;
					max_prob = p;
				}
			}
			var symbol = queues[best][0];
			var pos = buffer.length;
			
			for(var i = 0; i < n_symbols; i++) {
				var sym = this.probs[i*2];
				var prob = this.probs[i*2+1]<<8;
				var s = [((prob*symbol[0])>>>16), pos, symbol[2]+1]; //combine probs, keep track of buffer, keep length of queue

				for(var k  = 0; k < symbol[2]; k++)
					buffer[pos+k] = buffer[symbol[1] + k]; //copy sequence of symbols

				pos += symbol[2];
				buffer[pos++] = sym; //append symbol
				queues[i].push(s);
			}
			table_length += (n_symbols-1)*(symbol[2] + 1) +1; 
			n_words += n_symbols -1;
			queues[best].shift(); //remove first thing
		}

		this.index = new Uint32Array(n_words);
		this.lengths = new Uint32Array(n_words);
		this.table = new Uint8Array(table_length);
		var word = 0;
		var pos = 0;
		for(i = 0; i < queues.length; i++) {
			var queue = queues[i];
			for(var k = 0; k < queue.length; k++) {
				var s = queue[k];
				this.index[word] = pos;
				this.lengths[word] = s[2]; //length
				word++;

				for(var j = 0; j < s[2]; j++)
					this.table[pos + j] = buffer[s[1] + j]; //buffer of offset
				pos += s[2]; //length
			}
		}
	},
	createDecodingTables2: function() {
		//Here probs will range from 0 to 0xffff for better precision

		var t = this;
		var n_symbols = t.probs.length/2;
		if(n_symbols <= 1) return;

/*      QUEUE as column 

		A  x  AA
		C  C  CA
		G  G  GA
		T  T  TA
*/

		var dictionary_size = 1<<this.wordsize;

		//TODO: fill index and length: saves *3, and memory (compact in place, later)

		//(max length is word dictionary? nope the removed ones dont count, 1 per step so add max 255 potentially deleted)
		var queues = new Uint32Array(2*dictionary_size*3); //array of n symbols array of prob, position in buffer, length //limit to 8k*3
		var starts = new Uint32Array(n_symbols); //starts of each queue
		var end = 0; //keep track of queue end
		var buffer = t.table = new Uint8Array(dictionary_size*dictionary_size); //worst case for 2 symbols. 1 with 254 prob and other qwith 1 prob
		var pos = 0; //keep track of buffer first free space
		var table_length = 0;
		var n_words = 0;

		var count = 2;
		var p0 = (t.probs[1] << 8)>>>0; //first symbol
		var p1 = (t.probs[3]<<8)>>>0;    //second symbol.
		var prob = (p0*p0)>>>16;
		var max_count = (dictionary_size - 1)/(n_symbols - 1);
		while(prob > p1 && count < max_count) {
			prob = (prob*p0) >>> 16;
			count++;
		}

		if(count >= 16) { //Very low entropy results in large tables > 8K.
			buffer[pos++] = t.probs[0];
			for(var k = 1; k < n_symbols; k++) {
				for(var i = 0; i < count-1; i++)
					buffer[pos++] = t.probs[0];
				buffer[pos++] = t.probs[2*k];
			}
			starts[0] = (count-1)*n_symbols*3;
			for(var k = 1; k < n_symbols; k++)
				starts[k] = k*3;

			prob = 0xffff;
			for(var col = 0; col < count; col++) {
				for(var row = 1; row < n_symbols; row++) {
					var off = (row + col*n_symbols)*3;
					queues[off] = (prob * (t.probs[row*2+1]<<8)) >> 16;
					queues[off+1] = row*count - col;
					queues[off+1] = col+1;
				}
				prob = (prob*p0) >>> 16;
			}
			var first = ((count-1)*n_symbols)*3;
			queues[first] = prob;
			queues[first+1] = 0;
			queues[first+1] = count;
			n_words = 1 + count*(n_symbols - 1);
			table_length = n_words;
			end = count*n_symbols;

		} else {
			//initialize adding all symbols to queues
			for(var i = 0; i < n_symbols; i++) {
				var symbol = t.probs[i*2];
				var prob = t.probs[i*2+1];
				queues[i*3] = prob<<8;
				queues[i*3+1] = i;
				queues[i*3+2] = 1;
	
				starts[i] = i*3;
				buffer[i] = symbol;
			}
			pos = n_symbols;
			end = n_symbols*3;

			n_words = n_symbols;
			table_length = n_symbols;
		}

		//at each step we grow all queues using the most probable sequence
		while(n_words < dictionary_size - n_symbols +1) {
			//find highest probability word
			var best = 0;
			var max_prob = 0;
			for(var i = 0; i < n_symbols; i++) {
				var p = queues[starts[i]]; //front of queue probability.
				if(p > max_prob) {
					best = i;
					max_prob = p;
				}
			}
			var start = starts[best];
			var offset = queues[start+1];
			var len = queues[start+2];
			for(var i = 0; i < n_symbols; i++) {
				var sym = t.probs[i*2];
				var prob = t.probs[i*2+1]<<8;
				queues[end] = ((prob*queues[start])>>>16);
				queues[end+1] = pos;
				queues[end+2] = len + 1;
				end += 3;

				for(var k  = 0; k < len; k++)
					buffer[pos + k] = buffer[offset + k]; //copy sequence of symbols

				pos += len;
				buffer[pos++] = sym; //append symbol
			}
			starts[best] += n_symbols*3; //move one column

			table_length += (n_symbols-1)*(len + 1) +1;
			n_words += n_symbols -1;
		}

		t.index = new Uint32Array(n_words);
		t.lengths = new Uint32Array(n_words);
		var word = 0;
		var count = 0;
		for(i = 0, row = 0; i < end; i += 3, row++) {
			if(row >= n_symbols)
				row  = 0;
			if(starts[row] > i) continue; //skip deleted words

			t.index[word] = queues[i+1];
			t.lengths[word] = queues[i+2];
			word++;
		}
	},
	_decompress: function(input, input_size, output, output_size) {
		//TODO optimize using buffer arrays
		var input_pos = 0;
		var output_pos = 0;
		if(this.probs.length == 2) {
			var symbol = this.probs[0];
			for(var i = 0; i < output_size; i++)
				output[i] = symbol;
			return;
		}

		while(input_pos < input_size-1) {
			var symbol = input[input_pos++];
			var start = this.index[symbol];
			var end = start + this.lengths[symbol];
			for(var i = start; i < end; i++) 
				output[output_pos++] = this.table[i];
		}

		//last symbol might override so we check.
		var symbol = input[input_pos];
		var start = this.index[symbol];
		var end = start + output_size - output_pos;
		var length = output_size - output_pos;
		for(var i = start; i < end; i++)
			output[output_pos++] = this.table[i];

		return output;
	}
}

File: src/nxz/tunstall.cpp
/*
Nexus

Copyright(C) 2012 - Federico Ponchio
ISTI - Italian National Research Council - Visual Computing Lab

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License (http://www.gnu.org/licenses/gpl.txt)
for more details.
*/
#include <assert.h>
#include <math.h>
#include <string.h>
#include <deque>
#include <algorithm>
#include <iostream>
#include "tunstall.h"

using namespace std;
using namespace nx;

struct TSymbol {
	int offset;
	int length;
	uint32_t probability;
	TSymbol() {}
	TSymbol(uint32_t p, int o, int l): probability(p), offset(o), length(l) {}
};

/* Left as an example of how to compress to a stream

int Tunstall::compress(Stream &stream, unsigned char *data, int size) {

	Tunstall t;
	t.getProbabilities(data, size);

	t.createDecodingTables();
	t.createEncodingTables();

	int compressed_size;
	unsigned char *compressed_data = t.compress(data, size, compressed_size);

	stream.write<uchar>(t.probabilities.size());
	stream.writeArray<uchar>(t.probabilities.size()*2, (uchar *)&*t.probabilities.begin());


	stream.write<int>(size);
	stream.write<int>(compressed_size);
	stream.writeArray<unsigned char>(compressed_size, compressed_data);
	delete []compressed_data;
	//return compressed_size;
	return 1 + t.probabilities.size()*2 + 4 + 4 + compressed_size;
}

void Tunstall::decompress(Stream &stream, std::vector<unsigned char> &data) {

	Tunstall t;
	int nsymbols = stream.read<uchar>();
	uchar *probs = stream.readArray<uchar>(nsymbols*2);
	t.probabilities.resize(nsymbols);
	memcpy(&*t.probabilities.begin(), probs, nsymbols*2);

	t.createDecodingTables();

	int size = stream.read<int>();
	data.resize(size);
	int compressed_size = stream.read<int>();
	unsigned char *compressed_data = stream.readArray<unsigned char>(compressed_size);

	if(size)
		t.decompress(compressed_data, compressed_size, &*data.begin(), size);
} */


void Tunstall::getProbabilities(unsigned char *data, int size) {

#ifdef DEBUG_ENTROPY
	double e = 0;
#endif
	probabilities.clear();

	std::vector<int> probs(256, 0);
	for(int i = 0; i < size; i++)
		probs[data[i]]++;
//	int max = 0;
	for(size_t i = 0; i < probs.size(); i++) {
		if(probs[i] > 0) {
//			max = i;
			//TODO scaling probabilities to MAX would result in better accuracy (compression)?


#ifdef DEBUG_ENTROPY
			double p = probs[i]/(double)size;
			e -= p * log2(p);
			cout << (char)(i + 65) << " P: " << p << endl;
#endif
			probabilities.push_back(Symbol(i, probs[i]*255/size));
		}
	}
	std::sort(probabilities.begin(), probabilities.end(),
			  [](const Symbol &a, const Symbol &b)->bool { return a.probability > b.probability; });
#ifdef DEBUG_ENTROPY
	cout << "Entropy: " << e << " theorical compression: " << e*size/8 << endl;
#endif
}

void Tunstall::setProbabilities(float *probs, int n_symbols) {
	probabilities.clear();
	for(int i = 0; i < n_symbols; i++) {
		if(probs[i] <= 0) continue;
		probabilities.push_back(Symbol(i, probs[i]*255));
	}
}

void Tunstall::createDecodingTables2() {
	int n_symbols = probabilities.size();
	if(n_symbols <= 1) return;

	int dictionary_size = 1<<wordsize;
	std::vector<TSymbol> queues(2*dictionary_size);
	size_t end = 0;
	vector<unsigned char> &buffer = table;
	assert(wordsize == 8);
	buffer.resize(8192);
	//buffer.resize(65536);
	int pos = 0; //keep track of buffer lenght/
	vector<int> starts(n_symbols);

	int n_words = 0;

	int count = 2;
	uint32_t p0 = probabilities[0].probability<<8;
	uint32_t p1 = probabilities[1].probability<<8;
	uint32_t prob = (p0*p0)>>16;
	int max_count = (dictionary_size - 1)/(n_symbols - 1);
	while(prob > p1 && count < max_count) {
		prob = (prob*p0) >> 16;
		count++;
	}

	if(count >= 16) { //Very low entropy results in large tables > 8K.
		//  AAAA...A count times.    //only 1 word
		//  AAAA...B and we need all shorter to b so count
		//  AAAA...Z
		//nwords = (n_symbols-1)*count + 1 <= dictionary_size;
		//count <= (dictionary_size -1)/(n_symbols -1);
		//test 255/1 -> count = 255   A...A + 255 AAAAAB
		//test 255/2 -> count = 127   A...A + 127*A..B + 127*A...C = 255.

		buffer[pos++] = probabilities[0].symbol;
		for(int k = 1; k < n_symbols; k++) {
			for(int i = 0; i < count-1; i++)
				buffer[pos++] = probabilities[0].symbol;
			buffer[pos++] = probabilities[k].symbol;
		}
		starts[0] = (count-1)*n_symbols;
		for(int k = 1; k < n_symbols; k++)
			starts[k] = k;

		for(int col = 0; col < count; col++) {
			for(int row = 1; row < n_symbols; row++) {
				TSymbol &s = queues[row + col*n_symbols];
				if(col == 0)
					s.probability = (probabilities[row].probability<<8);
				else
					s.probability = (prob * (probabilities[row].probability<<8)) >> 16;
				s.offset = row*count - col;
				s.length = col+1;
			}
			if(col == 0)
				prob = p0;
			else
				prob = (prob*p0) >> 16;
		}
		TSymbol &first = queues[(count-1)*n_symbols];
		first.probability = prob;
		first.offset = 0;
		first.length = count;
		n_words = 1 + count*(n_symbols - 1);
		end = count*n_symbols;
		assert(n_words == pos);

	} else {
		n_words = n_symbols;
		//initialize adding all symbols to queues
		for(int i = 0; i < n_symbols; i++) {
			TSymbol s;
			s.probability = (uint32_t)(probabilities[i].probability<<8);
			s.offset = pos;
			s.length = 1;

			starts[i] = i;
			queues[end++] = s;
			buffer[pos++] = probabilities[i].symbol;
		}
	}

	while(n_words < dictionary_size - n_symbols +1) {
		//find highest probability word
		int best = 0;
		uint32_t max_prob = 0;
		for(int i = 0; i < n_symbols; i++) {
			uint32_t p = queues[starts[i]].probability ;
			if(p > max_prob) {
				best = i;
				max_prob = p;
			}
		}
		TSymbol &symbol = queues[starts[best]];

		for(int i = 0; i < n_symbols; i++) {
			uint32_t p = probabilities[i].probability;
			TSymbol s;
			s.probability = ( ( symbol.probability * (unsigned int)(p<<8) )>>16);
			s.offset = pos;
			s.length = symbol.length + 1;
			assert(pos + symbol.length < buffer.size());
			memcpy(&buffer[pos], &buffer[symbol.offset], symbol.length);
			pos += symbol.length;
			buffer[pos++] = probabilities[i].symbol;
			queues[end++] = s;
		}
		starts[best] += n_symbols;
		n_words += n_symbols -1;
	}
	index.clear();
	lengths.clear();

	//build table and index
	index.resize(n_words);
	lengths.resize(n_words);
	int word = 0;
	for(size_t i = 0, row = 0; i < end; i++, row++) {
		if(row >= n_symbols)
			row = 0;
		if(starts[row] > i)
			continue;

		TSymbol &s = queues[i];
		index[word] = s.offset;
		lengths[word] = s.length;
		word++;
	}
	//cout << "Table size: " << pos << endl;
}

void Tunstall::createDecodingTables() {
	int n_symbols = probabilities.size();
	if(n_symbols <= 1) return;

	vector<deque<TSymbol> > queues(n_symbols);
	vector<unsigned char> buffer;

	//initialize adding all symbols to queues
	for(int i = 0; i < n_symbols; i++) {
		TSymbol s;
		s.probability = (uint32_t)(probabilities[i].probability<<8);
		s.offset = buffer.size();
		s.length = 1;

		queues[i].push_back(s);
		buffer.push_back(probabilities[i].symbol);
	}
	int dictionary_size = 1<<wordsize;
	int n_words = n_symbols;
	int table_length = n_symbols;
	while(n_words < dictionary_size - n_symbols +1) {
		//find highest probability word
		int best = 0;
		float max_prob = 0;
		for(int i = 0; i < n_symbols; i++) {
			float p = queues[i].front().probability ;
			if(p > max_prob) {
				best = i;
				max_prob = p;
			}
		}

		TSymbol symbol = queues[best].front();
		//split word.
		int pos = buffer.size();
		buffer.resize(pos + n_symbols*(symbol.length + 1));
		for(int i = 0; i < n_symbols; i++) {
			uint32_t p = probabilities[i].probability;
			TSymbol s;
			//TODO check performances with +1 and without.
			s.probability = ( ( symbol.probability * (unsigned int)(p<<8) )>>16); //probabilities[i].probability*symbol.probability;
			s.offset = pos;
			s.length = symbol.length + 1;

			memcpy(&buffer[pos], &buffer[symbol.offset], symbol.length);
			pos += symbol.length;
			buffer[pos++] = probabilities[i].symbol;
			queues[i].push_back(s);
		}
		table_length += (n_symbols-1)*(symbol.length + 1) +1;
		n_words += n_symbols -1;
		queues[best].pop_front();
	}
	index.clear();
	lengths.clear();
	table.clear();

	//build table and index
	index.resize(n_words);
	lengths.resize(n_words);
	table.resize(table_length);
	int word = 0;
	int pos = 0;
	for(size_t i = 0; i < queues.size(); i++) {
		deque<TSymbol> &queue = queues[i];
		for(size_t k = 0; k < queue.size(); k++) {
			TSymbol &s = queue[k];
			index[word] = pos;
			lengths[word] = s.length;
			word++;
			memcpy(&table[pos], &buffer[s.offset], s.length);
			pos += s.length;
		}
	}
	assert((int)index.size() <= dictionary_size);
}

void Tunstall::createEncodingTables() {

	int n_symbols = probabilities.size();
	if(n_symbols <= 1) return; //not much to compress
	//we need to reverse the table and index
	int lookup_table_size = 1;
	for(int i = 0; i < lookup_size; i++)
		lookup_table_size *= n_symbols;

	remap.resize(256, 0);
	for(int i = 0; i < n_symbols; i++) {
		Symbol &s = probabilities[i];
		remap[s.symbol] = i;
	}

	if(probabilities[0].probability > rle_limit) return;


	offsets.clear();
	offsets.resize(lookup_table_size, 0xffffff); //this is enough for quite large tables.
	for(size_t i = 0; i < index.size(); i++) {
		int low, high;
		int offset = 0;
		int table_offset = 0;
		while(1) {
			wordCode(&table[index[i] + offset], lengths[i] - offset, low, high);
			if(lengths[i] - offset <= lookup_size) {
				for(int k = low; k < high; k++) {
					offsets[table_offset + k] = i;
				}
				break;
			}

			//word is too long
			//check if some other word already did this:
			if(offsets[table_offset + low] == 0xffffff) { //add
				offsets[table_offset + low] = -offsets.size();
				offsets.resize(offsets.size() + lookup_table_size, 0xffffff);
			}

			table_offset = -offsets[table_offset + low];
			offset += lookup_size;
		}
	}
}

unsigned char *Tunstall::compress(unsigned char *data, int input_size, int &output_size) {
	if(probabilities.size() == 1) {
		output_size = 0;
		return NULL;
	}

	unsigned char *output = new unsigned char[input_size*2]; //use entropy here!

	assert(wordsize <= 16);
	output_size = 0;
	int input_offset = 0;
	int word_offset = 0;
	int offset = 0;
	while(input_offset < input_size) {
		//read lookup_size symbols and compute code:
		int d = input_size - input_offset;
		if(d > lookup_size)
			d = lookup_size;
		int low, high;
		wordCode(data + input_offset, d, low, high);
		offset = offsets[-offset + low];
		assert(offset != 0xffffff);
		if(offset >= 0) { //ready to ouput a symbol
			output[output_size++] = offset;
			input_offset += lengths[offset] - word_offset;
			offset = 0;
			word_offset = 0;
		} else {
			word_offset += lookup_size;
			input_offset += lookup_size;
		}
	}
	//end of stream can be tricky:
	//we could have a partial read (need to encode half a word)
	if(offset < 0) {
		while(offset < 0)
			offset = offsets[-offset];
		output[output_size++] = offset;
	}
	assert(output_size <= input_size*2);
	return output;
}

void Tunstall::decompress(unsigned char *data, int input_size, unsigned char *output, int output_size) {
	unsigned char *end_output = output + output_size;
	unsigned char *end_data = data + input_size -1;
	if(probabilities.size() == 1) {
		memset(output, probabilities[0].symbol, output_size);
		return;
	}
	index.push_back(index.back() + lengths.back()); //TODO WHY?
	while(data < end_data) {
		int symbol = *data++;
		int start = index[symbol];
		int length = lengths[symbol];
		//int length = index[symbol+1] - start;
		memcpy(output, &table[start], length);
		output += length;
	}

	//last symbol might override so we check.
	int symbol = *data;
	int start = index[symbol];
	int length = (end_output - output);
	memcpy(output, &table[start], length);
}

int Tunstall::decompress(unsigned char *data, unsigned char *output, int output_size) {
	unsigned char *end_output = output + output_size;
	unsigned char *start = data;
	if(probabilities.size() == 1) {
		memset(output, probabilities[0].symbol, output_size);
		return 0;
	}
	exit(0);
	while(1) {
		int symbol = *data++;
		assert(symbol < (int)index.size());
		int start = index[symbol];
		int length = lengths[symbol];
		if(output + length >= end_output) {
			length = (end_output - output);
			memcpy(output, &table[start], length);
			break;
		} else {
			memcpy(output, &table[start], length);
			output += length;
		}
	}
	return data - start;
}


float Tunstall::entropy() {
	float e = 0;
	for(size_t i = 0; i < probabilities.size(); i++) {
		float p = probabilities[i].probability/255.0f;
		e += p*log(p)/log(2);
	}
	return -e;
}
